library/core/src/str/iter.rs

```diff
@@ -1497,3 +1498,346 @@ macro_rules! escape_types_impls {
 }
 
 escape_types_impls!(EscapeDebug, EscapeDefault, EscapeUnicode);
+
+/// Internal trait for Unicode conversions that potentially
+/// need context and can expand one char to several.
+///
+/// Default implementation is pass-through, no conversion is done,
+/// with `is_simple = is_ascii`.
+trait UnicodeConverter {
```

Comment:
I think this shouldn't use a trait. I think this ends up complicating things quite a bit and makes the implementation overly generic.